File: src/utils/topic_source/husky/ros_deep_learning/src/image_stitcher.py
#!/usr/bin/env python

import rospy
from sensor_msgs.msg import Image
from cv_bridge import CvBridge
import cv2
import numpy as np

class ImageStitcher:
    def __init__(self):
        self.bridge = CvBridge()
        self.image_sub1 = rospy.Subscriber("/camera1/image_raw", Image, self.image_callback1)
        self.image_sub2 = rospy.Subscriber("/camera2/image_raw", Image, self.image_callback2)
        self.image_sub3 = rospy.Subscriber("/camera3/image_raw", Image, self.image_callback3)
        self.image_pub = rospy.Publisher("/combined_image", Image, queue_size=10)

        self.image1 = None
        self.image2 = None
        self.image3 = None

    def image_callback1(self, msg):
        self.image1 = self.bridge.imgmsg_to_cv2(msg, "bgr8")
        self.stitch_images()

    def image_callback2(self, msg):
        self.image2 = self.bridge.imgmsg_to_cv2(msg, "bgr8")
        self.stitch_images()

    def image_callback3(self, msg):
        self.image3 = self.bridge.imgmsg_to_cv2(msg, "bgr8")
        self.stitch_images()

    def stitch_images(self):
        if self.image1 is not None and self.image2 is not None and self.image3 is not None:
            # Gabungkan gambar secara horizontal
            stitched_image = np.hstack((self.image1, self.image2, self.image3))
            stitched_image_msg = self.bridge.cv2_to_imgmsg(stitched_image, "bgr8")
            self.image_pub.publish(stitched_image_msg)

if __name__ == '__main__':
    rospy.init_node('image_stitcher')
    stitcher = ImageStitcher()
    rospy.spin()

File: src/utils/topic_source/husky/ros_deep_learning/src/image_stitcher.cpp
#include <ros/ros.h>
#include <sensor_msgs/Image.h>
#include <cv_bridge/cv_bridge.h>
#include <image_transport/image_transport.h>
#include <opencv2/opencv.hpp>
#include <opencv2/core/core.hpp>
#include <opencv2/imgproc/imgproc.hpp>

class ImageStitcher
{
public:
    ImageStitcher()
    {
        // Initialize ROS subscribers and publisher
        image_transport::ImageTransport it(nh_);
        image_sub1_ = it.subscribe("/camera1/image_raw", 1, &ImageStitcher::imageCallback1, this);
        image_sub2_ = it.subscribe("/camera2/image_raw", 1, &ImageStitcher::imageCallback2, this);
        image_sub3_ = it.subscribe("/camera3/image_raw", 1, &ImageStitcher::imageCallback3, this);
        image_pub_ = it.advertise("/combined_image", 1);

        // Initialize image pointers
        image1_ = nullptr;
        image2_ = nullptr;
        image3_ = nullptr;
    }

    // Callback untuk menerima gambar dari kamera 1
    void imageCallback1(const sensor_msgs::ImageConstPtr& msg)
    {
        try
        {
            image1_ = cv_bridge::toCvCopy(msg, "bgr8")->image;
            stitchImages();
        }
        catch (const cv_bridge::Exception& e)
        {
            ROS_ERROR("CV Bridge exception: %s", e.what());
        }
    }

    // Callback untuk menerima gambar dari kamera 2
    void imageCallback2(const sensor_msgs::ImageConstPtr& msg)
    {
        try
        {
            image2_ = cv_bridge::toCvCopy(msg, "bgr8")->image;
            stitchImages();
        }
        catch (const cv_bridge::Exception& e)
        {
            ROS_ERROR("CV Bridge exception: %s", e.what());
        }
    }

    // Callback untuk menerima gambar dari kamera 3
    void imageCallback3(const sensor_msgs::ImageConstPtr& msg)
    {
        try
        {
            image3_ = cv_bridge::toCvCopy(msg, "bgr8")->image;
            stitchImages();
        }
        catch (const cv_bridge::Exception& e)
        {
            ROS_ERROR("CV Bridge exception: %s", e.what());
        }
    }

    // Fungsi untuk menggabungkan gambar
    void stitchImages()
    {
        // Periksa jika semua gambar sudah diterima
        if (image1_ != nullptr && image2_ != nullptr && image3_ != nullptr)
        {
            // Gabungkan gambar secara horizontal
            cv::Mat stitched_image;
            cv::hconcat(image1_, image2_, stitched_image);
            cv::hconcat(stitched_image, image3_, stitched_image);

            // Konversi gambar OpenCV ke ROS message
            sensor_msgs::ImagePtr msg = cv_bridge::CvImage(std_msgs::Header(), "bgr8", stitched_image).toImageMsg();

            // Publikasikan gambar gabungan
            image_pub_.publish(msg);
        }
    }

private:
    ros::NodeHandle nh_;
    image_transport::Subscriber image_sub1_;
    image_transport::Subscriber image_sub2_;
    image_transport::Subscriber image_sub3_;
    image_transport::Publisher image_pub_;

    cv::Mat image1_;
    cv::Mat image2_;
    cv::Mat image3_;
};

int main(int argc, char** argv)
{
    // Inisialisasi ROS
    ros::init(argc, argv, "image_stitcher");
    ImageStitcher stitcher;
    ros::spin();
    return 0;
}
